core/trino-main/src/main/java/io/trino/sql/planner/CachingSubExpressionExtractor.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner;
+
+import io.trino.sql.tree.Expression;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+/**
+ * Extracts, caches and returns the stream of all expression subtrees within an Expression, including Expression itself
+ */
+public final class CachingSubExpressionExtractor
```

Comment:
This is only used in EqualityInference. Instead of having a whole dedicated class and abstraction for this (and for CachingSymbolsExtractor below), just inline the methods and the caches into EqualityInference.

Reply:
Not so easy to do it. `io.trino.sql.planner.EqualityInference#newInstance(io.trino.metadata.Metadata, java.util.Collection<io.trino.sql.tree.Expression>)` is a static, fabric method and uses both caches. If you inline caches to `EqualityInference` you need then `this` reference to access this caches. You do not have it in static, fabric method. I can move this complex stuff to constructor but on my eye it is not good idea (then, what is the point of fabric method here?). If you you really want to do it, let me know, I will do it.

Possibly, we can turn this seperate classes to internal static private classes, what do you think?

Reply:
You can solve that by moving all the logic into the constructor instead of having it split across the factory method and the constructor.

Reply:
@martint , done